.travis.yml

```diff
@@ -15,6 +15,9 @@ before_install: ./scripts/travis-install-build-deps.sh
 install: true
 
 script:
+  - export BRANCH=$(if [ "${TRAVIS_PULL_REQUEST}" == "false" ]; then echo ${TRAVIS_BRANCH}; else echo ${TRAVIS_PULL_REQUEST_BRANCH}-${TRAVIS_BRANCH}-pull-request; fi)
+  - git branch -D ${BRANCH} || true
+  - git checkout -b ${BRANCH} ${TRAVIS_COMMIT}
```

Comment:
Why not put the `|| true` at the end of this line, and delete line 19?

Reply:
My goal is to have the travis job run on the right commit (which it currently does), and with a sensible git branch name (which it currently doesn't).

When building pushes, the git branch is unspecified (detached HEAD) and Travis runs on the tip commit of the branch you pushed, as shown here: https://travis-ci.org/SebKuzminsky/pycam/jobs/292734192

When building PRs, github makes the merge on a throw-away integration branch and Travis runs on the merge commit, again with a detached HEAD, as shown here: https://travis-ci.org/SebKuzminsky/pycam/jobs/292734200

You can construct the desired branch name from Travis' environment variables (line 18).  Now, how do you convince git to use that desired branch name? 

If a branch exists and you "checkout -b" it, you get this error and HEAD doesn't move:
~~~~~
> cat .git/HEAD
ref: refs/heads/master
> git branch i-exist v0.6.0
> git checkout -b i-exist v0.6.1
fatal: A branch named 'i-exist' already exists.
> cat .git/HEAD
ref: refs/heads/master
~~~~~

I can't convince myself that it will always be the case in all Travis jobs that the branch name we want to build does *not* exist, but i *can* convince myself that the above code (lines 19 & 20) will always do what I want: remove the branch if it exists, then create it anew pointing to the commit i want.

Reply:
OK, that makes good sense.

I did not account for the failure mode where there's already a  branch ${BRANCH}, and it points at a different commit.